### PR TITLE
Record filters and masks

### PR DESCRIPTION
**Pull request by martint**
Adds a new field to query metadata section in the query completion event with the following structure:

```
"tables": [
   {
      "catalog": "...",
      "schema": "...",
      "table": "...",
      "authorization": "...",

      "filters": [ "...", ... ]
      "columns": [
         {
             "column": "..."
             "masks": [ "...", ...]
         },
         ...
      ]
   },
   ...
]
```

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/execution/Input.java [hunk 1]
```diff
@@ -30,15 +29,15 @@
 @Immutable
 public final class Input
 {
-    private final CatalogName catalogName;
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/execution/Input.java]: CatalogSchemaTableName or io.prestosql.metadata.QualifiedObjectName instead of using these three?

File: presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java [hunk 2]
```diff
@@ -148,6 +149,7 @@
 
     private final AtomicReference<Set<Input>> inputs = new AtomicReference<>(ImmutableSet.of());
     private final AtomicReference<Optional<Output>> output = new AtomicReference<>(Optional.empty());
+    private final AtomicReference<List<TableInfo>> referencedTables = new AtomicReference<>(ImmutableList.of());
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java]: use `null` initial values, to make sure it was not yet set, then in `setReferencedTables` please verify that previous value was `null`

File: presto-main/src/main/java/io/prestosql/sql/analyzer/Analysis.java [hunk 3]
```diff
@@ -481,17 +488,27 @@ public RelationType getOutputDescriptor(Node node)
 
     public TableHandle getTableHandle(Table table)
     {
-        return tables.get(NodeRef.of(table));
+        return tables.get(NodeRef.of(table)).getHandle().orElseThrow(() -> new IllegalArgumentException(format("%s is not a table reference", table)));
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/Analysis.java]:
```
return tables.get(NodeRef.of(table))
     .getHandle()
     .orElseThrow(() -> new IllegalArgumentException(format("%s is not a table reference", table)));
```

Please be polite for my poor eyesight.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/Analysis.java [hunk 4]
```diff
@@ -738,6 +760,39 @@ public void addColumnMask(Table table, String column, Expression mask)
...
+                    List<ColumnInfo> columns = referencedFields.get(table).stream()
+                            .map(field -> {
+                                String fieldName = field.getName().get();
+
+                                return new ColumnInfo(
+                                        fieldName,
+                                        columnMasks.getOrDefault(table, ImmutableMap.of())
+                                                .getOrDefault(fieldName, ImmutableList.of()).stream()
+                                                .map(Expression::toString)
+                                                .collect(Collectors.toList()));
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/Analysis.java]: toImmutableList()? static import. Apply the same for everywhere below.

File: presto-main/src/main/java/io/prestosql/dispatcher/FailedDispatchQuery.java [hunk 5]
```diff
@@ -226,6 +226,7 @@ private static QueryInfo immediateFailureQueryInfo(
                 ImmutableList.of(),
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/dispatcher/FailedDispatchQuery.java]: Do we have tests for event listener? Please add some.

File: presto-main/src/main/java/io/prestosql/dispatcher/FailedDispatchQuery.java [hunk 6]
```diff
@@ -226,6 +226,7 @@ private static QueryInfo immediateFailureQueryInfo(
...
                 ImmutableSet.of(),
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/dispatcher/FailedDispatchQuery.java]: There are missing information about accessed:
 - procedures (must have)
 - functions (must have)
 - DDL queries 
 - set session 
 - SHOW queries
 - query checks

All these surely should go as separate pull requests.

File: presto-main/src/main/java/io/prestosql/execution/Input.java (hunk 1 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/execution/Input.java]: I think that's overkill. QualifiedObjectName is not json-serializable, and CatalogSchemaTableName has a weird structure that just muddles things (it has catalog + schema-table as a nested object)

File: presto-main/src/main/java/io/prestosql/dispatcher/FailedDispatchQuery.java (hunk 6 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/dispatcher/FailedDispatchQuery.java]: DDL queries can be inferred from the query type + user + target table. Session properties can be inferred from the session properties in the query + user. Functions and procedures can be added, but I'll do it in a separate PR.

**Comment by martint** [comment]
Updated

File: presto-main/src/main/java/io/prestosql/sql/analyzer/Analysis.java [hunk 7]
```diff
@@ -738,6 +762,39 @@ public void addColumnMask(Table table, String column, Expression mask)
...
+                            info.getName().getCatalogName(),
+                            info.getName().getSchemaName(),
+                            info.getName().getObjectName(),
+                            info.getAuthorization(),
+                            rowFilters.getOrDefault(table, ImmutableList.of()).stream()
+                                    .map(Expression::toString)
+                                    .collect(toImmutableList()),
+                            columns);
+                })
+                .collect(Collectors.toList());
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/Analysis.java]: toImmutableList?

File: presto-main/src/main/java/io/prestosql/sql/analyzer/Analysis.java [hunk 8]
```diff
@@ -738,6 +762,39 @@ public void addColumnMask(Table table, String column, Expression mask)
...
+                                String fieldName = field.getName().get();
+
+                                return new ColumnInfo(
+                                        fieldName,
+                                        columnMasks.getOrDefault(table, ImmutableMap.of())
+                                                .getOrDefault(fieldName, ImmutableList.of()).stream()
+                                                .map(Expression::toString)
+                                                .collect(toImmutableList()));
+                            })
+                            .collect(Collectors.toList());
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/Analysis.java]: toImmutableList?

File: presto-main/src/main/java/io/prestosql/sql/analyzer/RelationId.java [hunk 9]
```diff
@@ -53,6 +53,12 @@ public boolean isAnonymous()
         return sourceNode == null;
     }
 
+    @Nullable
+    public Node getSourceNode()
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/RelationId.java]: Return `Optional<Node>`

File: presto-spi/src/main/java/io/prestosql/spi/eventlistener/TableInfo.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    private final String schema;
+    private final String table;
+    private final String authorization;
+
+    private final List<String> filters;
+    private final List<ColumnInfo> columns;
+
+    public TableInfo(String catalogName, String schema, String table, String authorization, List<String> filters, List<ColumnInfo> columns)
+    {
+        this.catalogName = catalogName;
```
Comment (kokosing) [inline presto-spi/src/main/java/io/prestosql/spi/eventlistener/TableInfo.java]: requireNonNulls

File: presto-spi/src/main/java/io/prestosql/spi/eventlistener/TableInfo.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        this.schema = schema;
+        this.table = table;
+        this.authorization = authorization;
+        this.filters = filters;
+        this.columns = columns;
```
Comment (kokosing) [inline presto-spi/src/main/java/io/prestosql/spi/eventlistener/TableInfo.java]: Please copy collections

File: presto-tests/src/test/java/io/prestosql/execution/TestEventListener.java [hunk 12]
```diff
@@ -250,6 +252,32 @@ public void testNormalQuery()
...
+        TableInfo table = tables.get(0);
+        assertEquals(table.getCatalogName(), "tpch");
+        assertEquals(table.getSchema(), "tiny");
+        assertEquals(table.getAuthorization(), "user");
+        assertTrue(table.getFilters().isEmpty());
+        assertEquals(table.getColumns().size(), 1);
+
+        ColumnInfo column = table.getColumns().get(0);
+        assertEquals(column.getColumn(), "linenumber");
+        assertTrue(column.getMasks().isEmpty());
```
Comment (kokosing) [inline presto-tests/src/test/java/io/prestosql/execution/TestEventListener.java]: it would be nice to see that filters and masks can be also collected. `io.prestosql.connector.MockConnectorFactory` could be modified to return some dummy masks and filters.

**Review by kokosing — APPROVED** [review]
% comments about tests

File: presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java (hunk 2 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java]: Actually, that doesn't work. There are many paths that don't even call the setInput/setOutput/setReferencedTables methods.

File: presto-tests/src/test/java/io/prestosql/execution/TestEventListener.java (hunk 12 above)
Comment (martint) [reply presto-tests/src/test/java/io/prestosql/execution/TestEventListener.java]: Masks and filters are part of AccessControl, not ConnectorMetadata, unfortunately. There's no infrastructure to replace the default access control in tests with a mock one. That would require some major surgery that I want to keep out of the scope of this PR.
